src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -1205,6 +1210,9 @@ private Task DestroyActivations(List<ActivationData> list)
                         directory.InvalidateCacheEntry(activationData.Address);
 
                         RerouteAllQueuedMessages(activationData, null, "Finished Destroy Activation");
+
+                        activationData.ServiceScope?.Dispose();
+                        activationData.ServiceScope = null;
```

Comment:
That's fair. The grain/activationdata creation flow is a little convoluted. It's more complicated to have some things internal and some things external, but it's alright.